packages/chain-adapters/src/utils/ledgerAppGate.ts

```diff
@@ -0,0 +1,126 @@
+import type { HDWallet } from '@shapeshiftoss/hdwallet-core'
+import { isLedger } from '@shapeshiftoss/hdwallet-ledger'
+import { KnownChainIds } from '@shapeshiftoss/types'
+import EventEmitter from 'events'
+
+export const emitter = new EventEmitter()
+
+export type LedgerOpenAppEventArgs = {
+  chainId: ChainId
+  isSigning: boolean
+  reject: (reason?: any) => void
+}
+
+export const getLedgerAppName = (chainId: ChainId | KnownChainIds | undefined) => {
+  switch (chainId as KnownChainIds) {
+    case KnownChainIds.ArbitrumMainnet:
+    case KnownChainIds.AvalancheMainnet:
+    case KnownChainIds.ArbitrumNovaMainnet:
+    case KnownChainIds.BaseMainnet:
+    case KnownChainIds.BnbSmartChainMainnet:
+    case KnownChainIds.EthereumMainnet:
+    case KnownChainIds.GnosisMainnet:
+    case KnownChainIds.OptimismMainnet:
+    case KnownChainIds.PolygonMainnet:
+      return 'Ethereum'
+    case KnownChainIds.BitcoinCashMainnet:
+      return 'Bitcoin Cash'
+    case KnownChainIds.BitcoinMainnet:
+      return 'Bitcoin'
+    case KnownChainIds.CosmosMainnet:
+      return 'Cosmos'
+    case KnownChainIds.DogecoinMainnet:
+      return 'Dogecoin'
+    case KnownChainIds.LitecoinMainnet:
+      return 'Litecoin'
+    case KnownChainIds.ThorchainMainnet:
+      return 'THORChain'
+    default:
+      throw Error(`Unsupported chainId: ${chainId}`)
+  }
+}
+
+const getCoin = (chainId: ChainId | KnownChainIds) => {
+  switch (chainId as KnownChainIds) {
+    case KnownChainIds.BitcoinMainnet:
+      return 'Bitcoin'
+    case KnownChainIds.DogecoinMainnet:
+      return 'Dogecoin'
+    case KnownChainIds.BitcoinCashMainnet:
+      return 'BitcoinCash'
+    case KnownChainIds.LitecoinMainnet:
+      return 'Litecoin'
+    case KnownChainIds.EthereumMainnet:
+      return 'Ethereum'
+    case KnownChainIds.AvalancheMainnet:
+      return 'Avalanche'
+    case KnownChainIds.OptimismMainnet:
+      return 'Optimism'
+    case KnownChainIds.BnbSmartChainMainnet:
+      return 'BnbSmartChain'
+    case KnownChainIds.PolygonMainnet:
+      return 'Polygon'
+    case KnownChainIds.GnosisMainnet:
+      return 'Gnosis'
+    case KnownChainIds.ArbitrumMainnet:
+      return 'Arbitrum'
+    case KnownChainIds.ArbitrumNovaMainnet:
+      return 'ArbitrumNova'
+    case KnownChainIds.BaseMainnet:
+      return 'Base'
+    case KnownChainIds.ThorchainMainnet:
+      return 'Rune'
+    case KnownChainIds.CosmosMainnet:
+      return 'Atom'
+    default:
+      throw Error(`Unsupported chainId: ${chainId}`)
+  }
+}
+
+export const verifyLedgerAppOpen = async (
+  chainId: ChainId | KnownChainIds,
+  wallet: HDWallet,
+  isSigning: boolean,
+) => {
+  const coin = getCoin(chainId)
+  const appName = getLedgerAppName(chainId)
+
+  if (!isLedger(wallet)) return
+
+  const isAppOpen = async () => {
+    try {
+      await wallet.validateCurrentApp(coin)
+      return true
+    } catch {
+      return false
+    }
+  }
+
+  if (await isAppOpen()) return
```

Comment:
How do we feel with this vs. keeping on polling in `useLedgerOpenApp`  in the previous implementation?

Feels like moving *all* logic to chain-adapters would miss opening/closing app scenarios as well as device locking, and the modal itself should still keep on checking for app open state on top of this

Reply:
I'm not sure I understand. We just keep polling in chain adapters until the app is opened before continuing with the function call. https://github.com/shapeshift/web/pull/7697/files/1194df7b3722f62718899d69880eb84ca99cb539#diff-bf20e794b650a4de95e6481d113dd7afcd65627ed5a7a1f23c33b07cc1acf1cfR105

Reply:
Actually disregard me - I was under the impression the current `useLedgerOpenApp` kept on polling regardless which it isn't - runtime behaviour should effectively be the same as before here!